### PR TITLE
Feature/new architecture proposal

### PR DESCRIPTION
**Pull request by marcellodesales**
This pull request is to improve and solve on Issue #15 ... In order to properly work, empty files should be passed to the stream `gulp.src()`. The README and Test cases were updated to reflect this proposal.

API Change
---------

Just changing the plugin to use streams and properly handle different protocols:
* file://
* http://
* https://
* File paths starting with "/"

```js
var withFilePath = {
  uri: process.env.PWD + "/package.json"
};
gulp.src([]).pipe(open(withFilePath));

// Open from the stream and default program
gulp.src("./package.json").pipe(open());

// Open from the stream using the app
var chrome = os.platform() === 'linux' ? 'google-chrome' : (
    os.platform() === 'darwin' ? 'google chrome' : (
    os.platform() === 'win32' ? 'chrome' : 'firefox'));
  gulp.src("./package.json").pipe(open({app: chrome}));
```

Output
-------

The task will show if a file or a URL was opened and if that was with the default or with the given app.

```
[19:04:56] Starting 'open'...
[19:04:56] Finished 'open' after 1.25 ms
[19:05:44] Opening the url http://localhost:3001/?ws=localhost:3001&port=5858 using the default OS app
[19:04:56] Opening the file /home/mdesales/dev/github-intuit/servicesplatform-nodejs/sp-reference-restify/package.json using the default OS app
```

Categorized Test Cases
------

```
$ npm test

> gulp-open@0.3.2 test /home/mdesales/dev/github/gulp-open
> mocha --reporter spec

  gulp-open
    opening files
      ✓ should open a stream of files with the default application
      ✓ should open a stream of files with a defined application
      ✓ should open a stream of files with a browser using the options
    opening urls
      ✓ should open a website in a browser using the options
      ✓ should open a website in a browser using chrome or firefox

  5 passing (5ms)
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,25 +1,104 @@
+var es = require('event-stream');
```
Comment (stevelacy) [inline index.js]: Use through2 as it is the gulp standard. Don't use es

File: index.js [hunk 2]
```diff
@@ -1,25 +1,104 @@
...
+var fs = require('fs');
```
Comment (stevelacy) [inline index.js]: Do not use any `fs` calls in any plugin

File: index.js [hunk 3]
```diff
@@ -1,25 +1,104 @@
...
-    if (file.isStream()) src = file.path;
-    if (!opt) opt = {};
-    var cmd = gutil.template(src, {file:file});
+var PLUGIN_NAME = 'gulp-open';
 
-    if (!opt.app && !opt.url) {
-      open(file.path);
-      return cb(null, file);
+// The acceptable uri formats for opening
+var ACCEPTABLE_URI_FORMATS = ["/", "file://", "http://", "https://"];
```
Comment (stevelacy) [inline index.js]: use single quotes

File: index.js [hunk 4]
```diff
@@ -1,25 +1,104 @@
...
+// Main plugin
+var gulpOpen = function(opt) {
+
+  var stream;
+  opt = opt || {};
+
+  // Executes the opening based on a given uri
+  var executeOpening = function() {
+    if (!opt.uri) {
+      stream.emit("error", new PluginError(PLUGIN_NAME, "URI is missing or incorrect! Use the src or the " +
```
Comment (stevelacy) [inline index.js]: single quotes

File: index.js [hunk 5]
```diff
@@ -1,25 +1,104 @@
...
+
+  // Executes the opening based on a given uri
+  var executeOpening = function() {
+    if (!opt.uri) {
+      stream.emit("error", new PluginError(PLUGIN_NAME, "URI is missing or incorrect! Use the src or the " + 
+        "options to indicate a uri " + ACCEPTABLE_URI_FORMATS));
+    }
+
+    // If the uri was not passed either by the stream nor by the options
+    var indexOfIndexOf = _indexOfIndexOf(opt.uri);
```
Comment (stevelacy) [inline index.js]: these names are odd: `indexOfIndexOf`,  `_indexOfIndexOf`

File: index.js [hunk 6]
```diff
@@ -1,25 +1,104 @@
...
+    if (indexOfIndexOf === -1) {
+      stream.emit("error", new PluginError(PLUGIN_NAME, "URI is missing or incorrect! Please use one of " + ACCEPTABLE_URI_FORMATS));
+    }
+
+    // If the type is the index value in the range of files
+    var selectedType = indexOfIndexOf <= 1 ? "file" : "url";
+    console.log(indexOfIndexOf);
```
Comment (stevelacy) [inline index.js]: do not use any `console.log`

File: index.js [hunk 7]
```diff
@@ -1,25 +1,104 @@
...
+
+    // If provided by the user using the options and not the stream, verify if the file exists first
+    if (selectedType === "file" && !opt.uriFromStream) {
+      try {
+        fs.statSync(opt.uri);
+
+      } catch (FileNotFoundError) {
```
Comment (stevelacy) [inline index.js]: do not use full camel case

File: index.js [hunk 8]
```diff
@@ -1,25 +1,104 @@
...
+    } else {
+      log(colors.blue('Opening the', selectedType, colors.green(opt.uri), 'using the app', colors.green(opt.app)));
+      open(opt.uri, opt.app);
     }
-    if(opt.url){
-      open(opt.url, opt.app);
-      return cb(null, file);
+  };
+
+  function done() {
```
Comment (stevelacy) [inline index.js]: this whole 'done' thing is fixed in through2

File: index.js [hunk 9]
```diff
@@ -1,25 +1,104 @@
...
-    // Run normally
-    open(cmd, opt.app);
-    cb(null, file);
-  });
-};
+  }
+
+  // Handles opening the file through the stream
+  var queueFile = function(file) {
+    console.log(file);
```
Comment (stevelacy) [inline index.js]: no logs

File: index.js [hunk 10]
```diff
@@ -1,25 +1,104 @@
...
+    // Just get the path and inject it to the options to be processed.
+    opt.uri = file.path;
+    opt.uriFromStream = true;
+  };
+
+  // Open the object, whatever that is.
+  var endStream = function() {
```
Comment (stevelacy) [inline index.js]: don't wrap unneeded functions

File: index.js [hunk 11]
```diff
@@ -1,25 +1,104 @@
...
+    executeOpening();
+  };
+
+  // copied from gulp-karma
+  stream = es.through(queueFile, endStream);
```
Comment (stevelacy) [inline index.js]: use through2.obj, not es.

File: package.json [hunk 12]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "gulp-open",
   "description": "Open files and URLs with gulp",
-  "version": "0.3.2",
+  "version": "1.0.0",
```
Comment (stevelacy) [inline package.json]: don't bump the version

File: test/main.js [hunk 13]
```diff
@@ -1,19 +1,35 @@
 var gulpopen = require('../');
 var should = require('should');
+var os = require("os");
```
Comment (stevelacy) [inline test/main.js]: use single quotes

**Comment by stevelacy** [comment]
Thanks for spending time writing this.
I commented on a few issues I noticed.

I forgot to add the correct linting file to the project as found here: https://github.com/stevelacy/gulp-stylus/blob/master/.jshintrc

There seems to be quite a bit of code added which could be simplified, first bt using through2, and second by having smaller functions for looping, etc.
There are some checks which are handled by gulp, if not then `open` handles it.

**Comment by marcellodesales** [comment]
@stevelacy Thanks for doing code review... Yeah, I did not use lint for the lack of it... Also, I did not know about the use of through2 vs es... I've seen this pattern in other modules before and I will have to read the docs from both in order to learn and use them properly...

So, before I spend any time doing this, do you think you would be able to merge the pull request in case I address all the required changes?

thanks a lot!
Marcello

**Comment by stevelacy** [comment]
Yes, I will merge it with the changes, provided that the examples work. They most likely will need to be updated to match your new API.

As a simple example of a through2 implementation: https://github.com/stevelacy/gulp-stylus/blob/master/index.js

emits are replaced with a simple return `cb(Error, Data)` pattern.
 
Thanks

File: index.js (hunk 5 above)
Comment (marcellodesales) [inline index.js]: Definitely :) Let me find something better....

Comment (marcellodesales) [inline index.js]: Is the following better?

```js
    // If the uri was not passed either by the stream nor by the options
    var indexOfAnyFormat = _getIndexesOfAcceptableFormats(opt.uri);
    if (indexOfAnyFormat === -1) {
      stream.emit('error', new PluginError(PLUGIN_NAME, 'URI is missing or incorrect! Please use one of ' + ACCEPTABLE_URI_FORMATS));
    }

    // If the type is the index value in the range of files
    var selectedType = indexOfAnyFormat <= 1 ? 'file' : 'url';
```

File: index.js (hunk 7 above)
Comment (marcellodesales) [inline index.js]: Block removed as I will not use "fs" to verify if the file exists.

File: index.js (hunk 9 above)
Comment (marcellodesales) [inline index.js]: Yeah completely missed that :(

File: package.json (hunk 12 above)
Comment (marcellodesales) [inline package.json]: Will revert that..

**Comment by marcellodesales** [comment]
@stevelacy Here's the first draft of the change... Can you please take a look again? 

I wrote this without seeing your message here :( I just reverted what you had done before and provided the updated samples in the README file, which was the original suggestion. 

Let me know if there's anything else I can do! Thanks a lot for the pointers!

**Comment by stevelacy** [comment]
Thanks for making those changes, I commented on two two items I noticed if you don't mind modifying, I'll then merge it.

**Comment by marcellodesales** [comment]
Hi @stevelacy... 

Thanks for another code review :+1: Yeah, I did not know what you would prefer, so I added the wget :) Just removed it... I also removed the reference to the array, and I'm using the suggested way for cb on errors.

**Pull request merged**